app/search/gleaner.py

```diff
@@ -28,48 +32,64 @@ def build_query(user_query="", page_number=1):
                 (GROUP_CONCAT(DISTINCT ?sameAs ; separator=", ") as ?sameAs)
                 (GROUP_CONCAT(DISTINCT ?keywords ; separator=", ") as ?keywords)
                 (GROUP_CONCAT(DISTINCT ?temporal_coverage ; separator=", ") as ?temporal_coverage)
+                (GROUP_CONCAT(DISTINCT ?spatial_coverage ; separator=", ") as ?spatial_coverage)
 
             {{
               VALUES ?type {{ schema:Dataset sschema:Dataset }}
               ?s a ?type .
               {{
                 ?s sschema:name ?title .
-                ?s sschema:keywords ?keywords .
-                ?s sschema:url ?url .
+                {{ ?s sschema:keywords ?keywords . }} UNION {{
+                    ?catalog ?relationship ?s .
+                    ?catalog sschema:keywords ?keywords .
+                }}
                 ?s sschema:description | sschema:description/sschema:value  ?abstract .
-                ?s sschema:identifier | sschema:identifier/sschema:value ?id .
                 ?s sschema:temporalCoverage ?temporal_coverage .
+                ?s sschema:spatialCoverage ?spatial_coverage .
 
                 OPTIONAL {{
                     ?s sschema:sameAs ?sameAs .
                 }}
                 OPTIONAL {{
                     ?s sschema:license ?license .
-                    
+                }}
+                OPTIONAL {{
+                    ?s sschema:url ?url .
+                }}
+                OPTIONAL {{
+                    ?s sschema:identifier | sschema:identifier/sschema:value ?id .
+                    FILTER(ISLITERAL(?id)) .
                 }}
               }}
               UNION
               {{
                 ?s schema:name ?title .
-                ?s schema:keywords ?keywords .
-                ?s schema:url ?url .
-                ?s schema:description | schema:description/schema:value ?abstract .
-                ?s schema:identifier | schema:identifier/schema:value ?id .
+                {{ ?s schema:keywords ?keywords . }} UNION {{
+                    ?catalog ?relationship ?s .
+                    ?catalog schema:keywords ?keywords .
+                }}
+                ?s schema:description | schema:description/schema:value  ?abstract .
                 ?s schema:temporalCoverage ?temporal_coverage .
+                ?s schema:spatialCoverage ?spatial_coverage .
 
                 OPTIONAL {{
                     ?s schema:sameAs ?sameAs .
                 }}
                 OPTIONAL {{
                     ?s schema:license ?license .
                 }}
-
+                OPTIONAL {{
+                    ?s schema:url ?url .
+                }}
+                OPTIONAL {{
+                    ?s schema:identifier | schema:identifier/schema:value ?id .
+                    FILTER(ISLITERAL(?id)) .
+                }}
               }}
-              FILTER(ISLITERAL(?id)) .
-
               {user_query}
+              BIND(COALESCE(?id, ?s) AS ?id)
```

Comment:
This is saying "if we didn't get something for `?id`, then set the RDF entity ID value as the id"